gateway-sp-comms/src/single_sp.rs

```diff
@@ -113,44 +111,65 @@ impl Drop for SingleSp {
 
 impl SingleSp {
     /// Construct a new `SingleSp` that will periodically attempt to discover an
-    /// SP reachable at `discovery_addr` (typically
-    /// [`DISCOVERY_MULTICAST_ADDR`], but possibly different in test /
-    /// development setups).
+    /// SP reachable on the port specified by `config`.
+    ///
+    /// This function returns immediately, but returns an object that is
+    /// initially (and possibly always) unusable: until we complete any local
+    /// setup for our UDP socket, methods of the returned `SingleSp` will fail.
+    /// The local setup includes:
+    ///
+    /// 1. Waiting for an interface with the name specified by
+    ///    `config.interface` to exist (if `config.interface` is `Some(_)`). We
+    ///    determine this via `if_nametoindex()`. This step never fails, but
+    ///    will block forever waiting for the interface.
+    /// 2. Binding a UDP socket to `config.listen_addr` (with a `scope_id`
+    ///    determined by the previous step). If this bind fails (e.g., because
+    ///    `config.listen_addr` is invalid), the returned `SingleSp` will return
+    ///    a "UDP bind failed" error from all methods forever.
```

Comment:
As discussed in chat, I think this is fine, because bind should only fail if the address is misconfigured or already in use. Both of those are invariants that should not be violated and require manual intervention to sort out.

The only other time bind would fail is if the interface returned somehow wasn't valid. This likely also requires manual intervention. It may be worth checking with @Nieuwejaar if this interface will get removed/recreated during the lifetime of a booted scrimlet.

Reply:
Under normal circumstances, I wouldn't expect it to.  However, if the sidecar gets powercycled or `dpd` or `tfportd` crash, then I would expect all of the `tfport` devices to be destroyed and recreated as part of the recovery.

Reply:
Oh boy. That could lead to interesting scenarios. Is it possible to have dpd or tfportd check to see if what it needs are already initialized, or must it recreate them? If the latter (or if that just makes things more rational form a viewpoint I'm not currently privy to), how would we go about detecting this. The interface ids could end up being wrong post bind, although maybe that's just a socket error that we retry on. I also wonder what happens if the ports try to get destroyed but because they are in use by the previously bound socket, illumos doesn't allow this. I'm not sure if this is possible or not, but it certainly seems like management here needs to be more robust to failures.

Reply:
Can we tie the services together so that if `dpd` or `tfportd` are restarted, `mgs` is too? (If we can, would that be a reasonable way to address this?)

Reply:
Ah great point! My Erlang [supervisor](https://www.erlang.org/doc/man/supervisor.html) brain should have thought of this. It looks like [SMF](https://illumos.org/man/7/smf) can do this via `restarters`.

Reply:
There probably are some specific cases where we could avoid rebuilding the world, but not in the general case.  We can't know how long we've been down and what may have changed in that time, so we can't know what should be cleaned up and what should be maintained.  We also can't know what information other daemons have about our state.

This has been a core design principle all along.  We have no mechanism for persisting/restoring any state at all between daemon launches.

Reply:
That makes sense. Thanks for the explanation Nils.

Reply:
In the event of the sidecar rebooting, I think our current plan is to restart the whole switch zone.

In the event of a daemon crash, If we think of `mgs` as being logically downstream of `dpd` and `tfportd`, then it makes sense to restart `mgs` when we restart the others.

Reply:
I should add that I don't have a firm opinion on whether it makes sense to think of `mgs` as downstream of `tfportd` or not.  Since `mgs` relies on interfaces that `tfportd` creates, then that feels like a reasonable mental model, but I don't know enough about `mgs` to say that with real certainty.

Reply:
While restarting everything is an ok short term approach, it isn't what we want to do long term as that's pretty disruptive and unfortunate for clients and consumers. Most other HTTP services would basically try to return that something downstream isn't working as opposed to just not existing. While we have a lot of gotchas for getting to that state, I think being there is going to be a healthier place for trying to reliably build applications on top of these things and dealing with them coming and going. I do understand that it's not that simple for us to do and won't be there for some time possibly.

To put this in perspective, nexus isn't going to restart itself just because it loses connectivity with cockroach. I realize that's not really a fair analogy, but just something to think here. I realize and appreciate that none of this is as simple as a remote HTTP service coming and going.